src/ripple/app/tx/impl/details/NFTokenUtils.cpp

```diff
@@ -520,34 +520,52 @@ findTokenAndPage(
     }
     return std::nullopt;
 }
-void
-removeAllTokenOffers(ApplyView& view, Keylet const& directory)
-{
-    view.dirDelete(directory, [&view](uint256 const& id) {
-        auto offer = view.peek(Keylet{ltNFTOKEN_OFFER, id});
 
-        if (!offer)
-            Throw<std::runtime_error>(
-                "Offer " + to_string(id) + " not found in ledger!");
+int
+removeTokenOffersWithLimit(
+    ApplyView& view,
+    Keylet const& directory,
+    int maxDeletableOffers)
+{
+    if (maxDeletableOffers == 0)
+        return 0;
 
-        auto const owner = (*offer)[sfOwner];
+    std::optional<std::uint64_t> pageIndex{0};
+    int deletedOffersCount = 0;
 
-        if (!view.dirRemove(
-                keylet::ownerDir(owner),
-                (*offer)[sfOwnerNode],
-                offer->key(),
-                false))
-            Throw<std::runtime_error>(
-                "Offer " + to_string(id) + " not found in owner directory!");
+    do
+    {
+        auto const page = view.peek(keylet::page(directory, *pageIndex));
+        if (!page)
+            break;
+
+        auto offerIndexes = page->getFieldV256(sfIndexes);
+
+        // We reverse-iterate the offer directory page to delete all entries.
+        // Deleting an entry in a NFTokenOffer directory page won't cause
+        // entries from other pages to move to the current, so, it is safe to
+        // delete entries one by one in the page. It is required to iterate
+        // backwards to handle iterator invalidation for vector, as we are
+        // deleting during iteration.
+        for (int i = offerIndexes.size() - 1; i >= 0; --i)
+        {
+            if (auto const offer = view.peek(keylet::nftoffer(offerIndexes[i])))
+            {
+                if (deleteTokenOffer(view, offer))
+                    ++deletedOffersCount;
+                else
+                    Throw<std::runtime_error>(
+                        "Offer " + to_string(offerIndexes[i]) +
+                        " cannot be deleted!");
+            }
 
-        adjustOwnerCount(
-            view,
-            view.peek(keylet::account(owner)),
-            -1,
-            beast::Journal{beast::Journal::getNullSink()});
+            if (maxDeletableOffers == deletedOffersCount)
+                break;
+        }
+        pageIndex = (*page)[~sfIndexNext];
```

Comment:
Yes, you didn't see a problem, but that still doesn't mean it was a good idea.

> Just to clarify, this situation is a problem because the `view.peek()` returns a shared_ptr right?

Yes, but it's better to take a larger view on this.  For every C++ value you can either have a copy or a reference.  The `shared_ptr` is a kind of reference.  Whenever you have a reference you can't really control when the thing referred to will go away.  In contrast, if you have a (local) copy, then you control all of the interactions with that value, including who has references to it.

That's a generalization.  There are specifics to lifetime rules using `shared_ptr` to `const` or non-`const` values.  But we don't need to explore those specifics here.

As it turns out, the `ApplyView` implementation saves up all of the deletes for the end of the transaction.  That's why you were still able to read `IndexNext` from the page node after the call to `erase`.  But it still was not a good idea.  After the call to `erase` on the node, what you had in your hands is called a "stale reference".  The value you want may still be there, but the implementation has every right to remove it due to some completely different software change.  What you had was a latent crash.  Thankfully that's fixed now.